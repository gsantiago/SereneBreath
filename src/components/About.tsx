import React from "react";
import { Popover } from "./Popover";

export function About({ isVisible }) {
  return (
    <Popover isVisible={isVisible} position="bottom">
      <section className="p-3">
        <p>made with love by @gsantiago</p>
      </section>
    </Popover>
  );
}
