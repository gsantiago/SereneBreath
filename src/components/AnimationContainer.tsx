import { animated, SpringValue, useSpringValue } from "@react-spring/web";

export const SIZE = 200;

export interface AnimationContainerProps {
  scale?: SpringValue<number>;
  children: React.ReactNode;
}

export function AnimationContainer({
  scale,
  children,
}: AnimationContainerProps) {
  const defaultScale = useSpringValue(1);

  return (
    <animated.div
      className="flex items-center justify-center "
      style={{ width: SIZE, height: SIZE, scale: scale ?? defaultScale }}
    >
      {children}
    </animated.div>
  );
}
