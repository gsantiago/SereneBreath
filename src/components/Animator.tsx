import React, { useEffect, useState } from "react";
import { AnimationProps } from "@/config/types";

export interface AnimatorProps
  extends Omit<AnimationProps, "state" | "isHolding"> {
  animation: React.FC<AnimationProps>;
}

export function Animator({
  animation: Animation,
  currentStep,
  pattern,
}: AnimatorProps) {
  const [state, setState] = useState<AnimationProps["state"]>("idle");

  const isHolding = currentStep === 1 || currentStep === 3;

  useEffect(() => {
    if (currentStep === 0 || currentStep === 1) {
      setState("inhaling");
    } else {
      setState("exhaling");
    }
  }, [currentStep]);

  return (
    <Animation
      currentStep={currentStep}
      pattern={pattern}
      state={state}
      isHolding={isHolding}
    />
  );
}
