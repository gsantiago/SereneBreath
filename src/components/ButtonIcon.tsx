import React from "react";

export interface ButtonIconProps
  extends Omit<React.HTMLProps<HTMLButtonElement>, "type"> {
  isActive?: boolean;
  position?: "top" | "bottom";
}

export function ButtonIcon({
  isActive,
  position = "top",
  ...props
}: ButtonIconProps) {
  const positionClass =
    position === "top" ? "rounded-t-md" : "rounded-b-md shadow-md";

  return (
    <button
      type="button"
      className={`px-2 pb-2 pt-2 text-3xl dark:text-white ${
        isActive
          ? `z-20 bg-white dark:bg-slate-800 ${positionClass}`
          : "rounded-md hover:bg-white hover:shadow-sm dark:hover:bg-slate-800"
      }`}
      {...props}
    />
  );
}
