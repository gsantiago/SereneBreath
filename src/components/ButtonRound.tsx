import React, { HTMLAttributes } from "react";

export function ButtonRound({
  style,
  ...props
}: HTMLAttributes<HTMLButtonElement>) {
  return (
    <button
      className="flex h-12 w-12 items-center justify-center rounded-full bg-teal-500 text-2xl text-white hover:bg-teal-600"
      {...props}
    />
  );
}
