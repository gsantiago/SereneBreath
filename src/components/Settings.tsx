import { Field } from "@/components/Field";
import { Stepper } from "@/components/Stepper";
import { RadioGroup } from "@/components/RadioGroup";
import { Switch } from "@/components/Switch";
import { Toggle } from "@/components/Toggle";
import { Popover } from "@/components/Popover";
import { TechniqueSelect } from "@/components/TechniqueSelect";

import { useStorage } from "@/hooks/useStorage";
import { useTranslation } from "@/hooks/useTranslation";

import { Settings as SettingsType } from "@/config/types";

import { canVibrate } from "@/modules/vibrator";

export interface SettingsProps {
  isVisible: boolean;
}

export function Settings({ isVisible }: SettingsProps) {
  const { t } = useTranslation();

  const [time, setTime] = useStorage("time");
  const [vibration, setVibration] = useStorage("vibration");
  const [guide, setGuide] = useStorage("guide");
  const [theme, setTheme] = useStorage("theme");

  return (
    <Popover isVisible={isVisible}>
      <TechniqueSelect />
      <Field
        id="time"
        label={t("settings.time.title")}
        description={t("settings.time.description")}
      >
        <Stepper min={1} value={time} onChange={setTime} />
      </Field>
      {canVibrate && (
        <Field
          id="vibration"
          label={t("settings.vibration.title")}
          description={t("settings.vibration.description")}
        >
          <Switch id="vibration" value={vibration} onChange={setVibration} />
        </Field>
      )}
      <Field
        id="guide"
        label={t("settings.guide.title")}
        description={
          guide === "disabled"
            ? t("settings.guide.disabled")
            : t("settings.guide.enabled")
        }
        children={
          <Switch
            id="guide"
            value={guide !== "disabled"}
            onChange={(selected) => setGuide(selected ? "female" : "disabled")}
          />
        }
        bottom={
          guide !== "disabled" && (
            <RadioGroup<SettingsType["guide"]>
              name="guide"
              value={guide}
              onChange={setGuide}
              options={[
                { label: t("settings.guide.options.female"), value: "female" },
                { label: t("settings.guide.options.male"), value: "male" },
                { label: t("settings.guide.options.bell"), value: "bell" },
              ]}
            />
          )
        }
      />
      <Field
        id="theme"
        label={t("settings.theme.title")}
        bottom={
          <Toggle<SettingsType["theme"]>
            id="theme"
            name="theme"
            value={theme}
            onChange={setTheme}
            options={[
              {
                label: t("settings.theme.system"),
                value: "system",
              },
              {
                label: t("settings.theme.dark"),
                value: "dark",
              },
              {
                label: t("settings.theme.light"),
                value: "light",
              },
            ]}
          />
        }
      />
    </Popover>
  );
}
