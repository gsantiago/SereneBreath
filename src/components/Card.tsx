import { RingAnimation } from "@/animations/RingAnimation";
import { Animator } from "@/components/Animator";
import { Pattern } from "@/config/types";

export const WIDTH = 250;
const HEIGHT = "100%";

export interface CardProps {
  title: string;
  description: string;
  pattern: Pattern;
}

export function Card({ title, description, pattern }: CardProps) {
  return (
    <section
      className="overflow-hidden rounded-lg bg-white"
      style={{ width: WIDTH, height: HEIGHT, paddingBottom: 45 }}
    >
      <div className="flex h-1/2 w-full items-center justify-center bg-orange-100">
        <div style={{ transform: "scale(0.8)" }}>
          <Animator
            currentStep={3}
            pattern={[2, 2, 2, 2]}
            animation={RingAnimation}
          />
        </div>
      </div>
      <div className="p-5">
        <h1 className="text-xl font-bold sm:text-2xl">{title}</h1>
        <p className="my-1 text-xs text-gray-400">{pattern.join("-")}</p>
        <p className="text-sm text-gray-800 sm:text-base">{description}</p>
      </div>
    </section>
  );
}
