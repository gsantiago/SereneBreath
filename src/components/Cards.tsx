import { Carousel } from "@/components/Carousel";
import { Card, WIDTH as CARD_WIDTH } from "@/components/Card";

import { useTranslation } from "@/hooks/useTranslation";
import { useStorage } from "@/hooks/useStorage";

import { techniques } from "@/config/techniques";
import { Technique } from "@/config/types";

export function Cards() {
  const { t } = useTranslation();

  const [technique, setTechnique] = useStorage("technique");
  const activeIndex = techniques.findIndex((t) => t.name === technique);

  return (
    <Carousel<Technique>
      activeIndex={activeIndex}
      itemWidth={CARD_WIDTH}
      items={[...techniques]}
      onChange={(item) => setTechnique(item.name)}
    >
      {({ item, isActive }) => (
        <Card
          title={t(`techniques.${item.name}.title`)}
          description={t(`techniques.${item.name}.description`)}
          pattern={item.pattern}
          isActive={isActive}
        />
      )}
    </Carousel>
  );
}
