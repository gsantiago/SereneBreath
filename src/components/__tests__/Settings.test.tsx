import { render, axe, screen, fireEvent } from "@/test-utils";
import { Settings, SettingsProps } from "@/components/Settings";

const setup = (props?: Partial<SettingsProps>) =>
  render(<Settings isVisible {...props} />);

test("renders successfuly", () => {
  const { container } = setup();
  expect(container).toMatchSnapshot();
});

test("accessibility", async () => {
  const { container } = setup();
  expect(await axe(container)).toHaveNoViolations();
});
