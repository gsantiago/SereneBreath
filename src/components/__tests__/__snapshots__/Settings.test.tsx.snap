// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders successfuly 1`] = `
<div>
  <div
    class="absolute  w-full rounded-md  bg-white py-2 shadow-md dark:bg-slate-800 sm:w-80  right-0 top-full rounded-tr-none z-10"
  >
    <div
      class="px-5 py-2"
    >
      <div
        class="grid grid-cols-6 items-end"
      >
        <div
          class="col-span-4"
        >
          <label
            class="text-sm font-medium text-gray-800 dark:text-white"
            for="guide"
          >
            Guide Breath
          </label>
          <p
            class="text-xs text-gray-500 dark:text-gray-400"
          >
            Select your favorite guide
          </p>
        </div>
        <div
          class="col-span-2 text-right"
        >
          <div
            class="relative inline-flex cursor-pointer items-center"
          >
            <input
              checked=""
              class="peer sr-only"
              id="guide"
              type="checkbox"
            />
            <button
              class="peer h-6 w-11 rounded-full bg-gray-200 after:absolute after:top-[2px] after:left-[2px] after:h-5 after:w-5 after:rounded-full after:border after:border-gray-300 after:bg-white after:transition-all after:content-[''] peer-checked:bg-blue-600 peer-checked:after:translate-x-full peer-checked:after:border-white peer-focus:outline-none peer-focus:ring-4 peer-focus:ring-blue-300 dark:border-gray-600 dark:bg-gray-700 dark:peer-focus:ring-blue-800"
              title="Switch"
            />
          </div>
        </div>
      </div>
      <div
        class="mt-2"
      >
        <div
          class="flex items-center"
        >
          <input
            checked=""
            class="h-4 w-4 cursor-pointer border-gray-300 text-blue-600 focus:ring-blue-600"
            id="option-guide-female"
            name="guide"
            type="radio"
            value="female"
          />
          <label
            class="text-md block cursor-pointer pl-2 text-xs leading-6 text-gray-700 dark:text-gray-100"
            for="option-guide-female"
          >
            Female voice
          </label>
        </div>
        <div
          class="flex items-center"
        >
          <input
            class="h-4 w-4 cursor-pointer border-gray-300 text-blue-600 focus:ring-blue-600"
            id="option-guide-male"
            name="guide"
            type="radio"
            value="male"
          />
          <label
            class="text-md block cursor-pointer pl-2 text-xs leading-6 text-gray-700 dark:text-gray-100"
            for="option-guide-male"
          >
            Male voice
          </label>
        </div>
        <div
          class="flex items-center"
        >
          <input
            class="h-4 w-4 cursor-pointer border-gray-300 text-blue-600 focus:ring-blue-600"
            id="option-guide-bell"
            name="guide"
            type="radio"
            value="bell"
          />
          <label
            class="text-md block cursor-pointer pl-2 text-xs leading-6 text-gray-700 dark:text-gray-100"
            for="option-guide-bell"
          >
            Bell
          </label>
        </div>
      </div>
    </div>
    <div
      class="px-5 py-2"
    >
      <div
        class="grid grid-cols-6 items-end"
      >
        <div
          class="col-span-4"
        >
          <label
            class="text-sm font-medium text-gray-800 dark:text-white"
            for="theme"
          >
            Theme
          </label>
        </div>
        <div
          class="col-span-2 text-right"
        />
      </div>
      <div
        class="mt-2"
      >
        <div
          class="flex"
        >
          <label
            class="cursor-pointer border  py-2 px-3 text-xs first:rounded-tl-md first:rounded-bl-md first:border-r-0 last:rounded-tr-md last:rounded-br-md last:border-l-0 hover:border-blue-600 hover:bg-blue-600 hover:text-white border-blue-600 bg-blue-600 text-white"
          >
            <input
              checked=""
              class="sr-only"
              name="theme"
              type="radio"
              value="system"
            />
            System
          </label>
          <label
            class="cursor-pointer border  py-2 px-3 text-xs first:rounded-tl-md first:rounded-bl-md first:border-r-0 last:rounded-tr-md last:rounded-br-md last:border-l-0 hover:border-blue-600 hover:bg-blue-600 hover:text-white border-gray-300 bg-gray-50"
          >
            <input
              class="sr-only"
              name="theme"
              type="radio"
              value="dark"
            />
            Dark
          </label>
          <label
            class="cursor-pointer border  py-2 px-3 text-xs first:rounded-tl-md first:rounded-bl-md first:border-r-0 last:rounded-tr-md last:rounded-br-md last:border-l-0 hover:border-blue-600 hover:bg-blue-600 hover:text-white border-gray-300 bg-gray-50"
          >
            <input
              class="sr-only"
              name="theme"
              type="radio"
              value="light"
            />
            Light
          </label>
        </div>
      </div>
    </div>
    <div
      class="px-5 py-2"
    >
      <div
        class="grid grid-cols-6 items-end"
      >
        <div
          class="col-span-4"
        >
          <label
            class="text-sm font-medium text-gray-800 dark:text-white"
            for="locale"
          >
            Language
          </label>
        </div>
        <div
          class="col-span-2 text-right"
        />
      </div>
      <div
        class="mt-2"
      >
        <select
          class="w-full cursor-pointer rounded-md border-gray-300 text-sm dark:text-gray-700"
          id="locale"
        >
          <option
            value="en"
          >
            English
          </option>
          <option
            value="pt"
          >
            PortuguÃªs
          </option>
        </select>
      </div>
    </div>
    <div
      class="px-5 py-1 dark:text-white"
    >
      <p
        class="mt-2 text-xs"
      >
        SereneBreath.com is a small webapp to help you improve your breathing, relaxing and managing your stress.
      </p>
      <p
        class="mt-2 text-xs"
      >
        This project is open source and open for new contributions. Check out its repo for more info:
         
        <a
          class="text-blue-600 underline dark:text-sky-300"
          href="https://github.com/gsantiago/SereneBreath"
          target="_blank"
        >
          Github Repo.
        </a>
      </p>
      <p
        class="mt-2 text-xs"
      >
        English and bell audio files are from the great Breathly app, which also inspired this app.
         
        <a
          class="text-blue-600 underline dark:text-sky-300"
          href="https://breathly.app/"
          target="_blank"
        >
          Breathly Site.
        </a>
      </p>
      <div
        class="mt-5 text-center"
      >
        <small
          class="text-xs text-gray-400"
        >
          v
          0.0.1
        </small>
      </div>
    </div>
  </div>
</div>
`;
